cli/src/commands/parse.rs

```diff
@@ -0,0 +1,53 @@
+//! `phylum parse` command for lockfile parsing
+
+use std::io;
+use std::path::Path;
+
+use anyhow::Result;
+use phylum_types::types::package::{PackageDescriptor, PackageType};
+
+use super::lock_files;
+use crate::lockfiles::*;
+
+type ParserResult = Result<(Vec<PackageDescriptor>, PackageType)>;
+
+const LOCKFILE_PARSERS: &[(&str, &dyn Fn(&Path) -> ParserResult)] = &[
+    ("yarn", &lock_files::parse::<YarnLock>),
+    ("npm", &lock_files::parse::<PackageLock>),
+    ("gem", &lock_files::parse::<GemLock>),
+    ("pip", &lock_files::parse::<PyRequirements>),
+    ("pipenv", &lock_files::parse::<PipFile>),
+    ("poetry", &lock_files::parse::<Poetry>),
+    ("mvn", &lock_files::parse::<Pom>),
+    ("gradle", &lock_files::parse::<GradleDeps>),
+    ("nuget", &lock_files::parse::<CSProj>),
+    ("auto", &lock_files::get_packages_from_lockfile),
+];
```

Comment:
Big fan of the dynamic dispatch of the referenced functions here. That does make things pretty clean and I tried doing something similar in the `lock_files.rs` cleanup but failed since the `Parser::new` function cannot be used easily due to different return types.

But now that the `parse` function exists already anyway, I wonder if this same method could be used to get rid of the macro in `try_get_packages`? It seems like it should be possible.

Reply:
Hmmm.... So we would avoid the need for a macro (which you [previously mentioned](https://github.com/phylum-dev/cli/pull/354/files#r867137035)) by using a loop instead?

That would make sense.

Reply:
The other option that I have considered for cleaning all of this up is to use [the strategy pattern](https://rust-unofficial.github.io/patterns/patterns/behavioural/strategy.html) and use unit types to implement the `Parseable` trait.